File: frontend/src/api/tasks.ts
import { api } from './client'

export interface TaskStep {
  id: number
  task_id: number
  step_key: string
  status: string
  input_snapshot: string | null
  output_snapshot: string | null
  error_message: string | null
  created_at: string
  updated_at: string
}

export interface TaskDetail {
  id: number
  user_id: string | null
  name: string | null
  status: string
  created_at: string
  updated_at: string
  profile_id?: number | null
  profile_name?: string | null
  steps: TaskStep[]
}

export interface CreateTaskResponse {
  id: number
  name: string | null
  status: string
  created_at: string
  profile_id?: number | null
  profile_name?: string | null
}

export interface TaskCompareSummary {
  has_framework: boolean
  chapter_count: number
}

export interface TaskSummary {
  id: number
  name: string | null
  status: string
  created_at: string
  profile_id?: number | null
  profile_name?: string | null
  compare_summary?: TaskCompareSummary | null
}

export async function createTask(args?: {
  initialSteps?: string[]
  name?: string
  profileId?: number | null
}): Promise<CreateTaskResponse> {
  const { data } = await api.post<CreateTaskResponse>('/api/tasks', {
    initial_steps: args?.initialSteps ?? undefined,
    name: args?.name ?? undefined,
    profile_id: args?.profileId ?? undefined,
  })
  return data
}

export async function getTask(id: string): Promise<TaskDetail> {
  const { data } = await api.get<TaskDetail>(`/api/tasks/${id}`)
  return data
}

/**
 * @param profileFilter - `null`: 仅内置默认（未绑定 Profile）的任务；`number`: 仅该 PromptProfile；`undefined`: 不传参，返回全部（兼容旧调用）
 */
export async function getTasks(profileFilter?: number | null): Promise<TaskSummary[]> {
  const params: Record<string, string> = {}
  if (profileFilter === null) {
    params.profile_id = 'default'
  } else if (profileFilter !== undefined && typeof profileFilter === 'number') {
    params.profile_id = String(profileFilter)
  }
  const { data } = await api.get<TaskSummary[]>('/api/tasks', {
    params: Object.keys(params).length > 0 ? params : undefined,
  })
  return data
}

export async function deleteTask(id: string): Promise<void> {
  await api.delete(`/api/tasks/${id}`)
}

export interface CancelTaskResponse {
  message: string
  revoked: boolean
  step_key?: string
}

export async function cancelTask(taskId: string): Promise<CancelTaskResponse> {
  const { data } = await api.post<CancelTaskResponse>(`/api/tasks/${taskId}/cancel`)
  return data
}

export interface UploadTaskFileResponse {
  step_key: string
  status: string
  message?: string
  stored_path?: string
}

export async function uploadTaskFile(taskId: string, file: File): Promise<UploadTaskFileResponse> {
  const formData = new FormData()
  formData.append('file', file)
  const { data } = await api.post<UploadTaskFileResponse>(`/api/tasks/${taskId}/upload`, formData)
  return data
}

export interface RunExtractStepResponse {
  message: string
  step_key: string
}

export async function runExtractStep(taskId: string): Promise<RunExtractStepResponse> {
  const { data } = await api.post<RunExtractStepResponse>(`/api/tasks/${taskId}/steps/extract/run`)
  return data
}

export interface RunAnalyzeStepResponse {
  message: string
  step_key: string
}

export async function runAnalyzeStep(taskId: string): Promise<RunAnalyzeStepResponse> {
  const { data } = await api.post<RunAnalyzeStepResponse>(`/api/tasks/${taskId}/steps/analyze/run`)
  return data
}

export interface RunParamsStepResponse {
  message: string
  step_key: string
}

export async function runParamsStep(taskId: string): Promise<RunParamsStepResponse> {
  const { data } = await api.post<RunParamsStepResponse>(`/api/tasks/${taskId}/steps/params/run`)
  return data
}

export interface RunFrameworkStepResponse {
  message: string
  step_key: string
}

export async function runFrameworkStep(taskId: string): Promise<RunFrameworkStepResponse> {
  const { data } = await api.post<RunFrameworkStepResponse>(`/api/tasks/${taskId}/steps/framework/run`)
  return data
}

export async function regenerateFrameworkStep(taskId: string): Promise<RunFrameworkStepResponse> {
  const { data } = await api.post<RunFrameworkStepResponse>(`/api/tasks/${taskId}/steps/framework/regenerate`)
  return data
}

export interface AcceptFrameworkStepResponse {
  message: string
  step_key: string
}

export async function saveFrameworkPoints(
  taskId: string,
  added_points: string[],
): Promise<AcceptFrameworkStepResponse> {
  const { data } = await api.post<AcceptFrameworkStepResponse>(
    `/api/tasks/${taskId}/steps/framework/save-points`,
    { added_points },
  )
  return data
}

export async function acceptFrameworkStep(
  taskId: string,
  added_points: string[],
): Promise<AcceptFrameworkStepResponse> {
  const { data } = await api.post<AcceptFrameworkStepResponse>(
    `/api/tasks/${taskId}/steps/framework/accept`,
    { added_points },
  )
  return data
}

export interface RunChaptersStepResponse {
  message: string
  step_key: string
}

export async function runChaptersStep(
  taskId: string,
  chapterNumbers?: number[],
): Promise<RunChaptersStepResponse> {
  const { data } = await api.post<RunChaptersStepResponse>(
    `/api/tasks/${taskId}/steps/chapters/run`,
    chapterNumbers != null && chapterNumbers.length > 0 ? { chapter_numbers: chapterNumbers } : {},
  )
  return data
}

export interface SaveChapterPointsResponse {
  message: string
  step_key: string
}

export async function saveChapterPoints(
  taskId: string,
  chapterNumber: number,
  addedPoints: string[],
): Promise<SaveChapterPointsResponse> {
  const { data } = await api.post<SaveChapterPointsResponse>(
    `/api/tasks/${taskId}/steps/chapters/save-points`,
    { chapter_number: chapterNumber, added_points: addedPoints },
  )
  return data
}

export interface RegenerateChapterResponse {
  message: string
  step_key: string
}

export async function regenerateChapter(
  taskId: string,
  chapterNumber: number,
  addedPoints?: string[],
): Promise<RegenerateChapterResponse> {
  const { data } = await api.post<RegenerateChapterResponse>(
    `/api/tasks/${taskId}/steps/chapters/regenerate`,
    addedPoints != null && addedPoints.length > 0
      ? { chapter_number: chapterNumber, added_points: addedPoints }
      : { chapter_number: chapterNumber },
  )
  return data
}

export interface RunReviewStepResponse {
  message: string
  step_key: string
}

export async function runReview(
  taskId: string,
  chapterNumber?: number,
): Promise<RunReviewStepResponse> {
  const url =
    chapterNumber != null
      ? `/api/tasks/${taskId}/steps/review/run?chapter_number=${chapterNumber}`
      : `/api/tasks/${taskId}/steps/review/run`
  const { data } = await api.post<RunReviewStepResponse>(url)
  return data
}

export interface AcceptReviewResponse {
  message: string
  step_key: string
}

export async function acceptReview(
  taskId: string,
  chapterNumber: number,
  acceptedItems: string[],
): Promise<AcceptReviewResponse> {
  const { data } = await api.post<AcceptReviewResponse>(
    `/api/tasks/${taskId}/steps/review/accept`,
    { chapter_number: chapterNumber, accepted_items: acceptedItems },
  )
  return data
}

export interface RegenerateAllFromReviewResponse {
  message: string
  step_key: string
}

export async function regenerateAllChaptersFromReview(
  taskId: string,
): Promise<RegenerateAllFromReviewResponse> {
  const { data } = await api.post<RegenerateAllFromReviewResponse>(
    `/api/tasks/${taskId}/steps/review/regenerate-all`,
  )
  return data
}

/** Download task DOCX. Returns blob for programmatic download. */
export async function downloadTaskDocx(taskId: string): Promise<Blob> {
  const { data } = await api.get(`/api/tasks/${taskId}/download`, {
    responseType: 'blob',
  })
  return data
}

File: backend/app/routers/tasks.py
"""Task CRUD, list and cancel API.

All step-specific endpoints have been extracted to dedicated router modules:
  upload.py    — file upload
  steps.py     — extract / analyze / params triggers
  framework.py — framework run / regenerate / save-points / accept / diff
  chapters.py  — chapters run / save-points / regenerate / regenerate-all / diff
  review.py    — review run / accept
  export.py    — DOCX download
"""
import logging
import shutil
from datetime import datetime

from fastapi import APIRouter, Body, Depends
from sqlalchemy.orm import Session

from app import config
from app.database import get_db
from app.models import Task, TaskStep
from app.schemas.compare import ChapterCompareMetaItem, CompareMetaResponse, FrameworkCompareMeta
from app.schemas.task import (
    CreateTaskRequest,
    CreateTaskResponse,
    DEFAULT_INITIAL_STEPS,
    TaskCompareSummary,
    TaskDetailResponse,
    TaskStepSchema,
    TaskSummary,
)
from app.services.step_service import (
    compute_compare_meta_for_task,
    compute_compare_meta_from_steps,
    require_task,
)
from celery_app import app as celery_app

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/tasks", tags=["tasks"])


@router.post("", response_model=CreateTaskResponse, status_code=201)
def create_task(
    body: CreateTaskRequest | None = Body(None),
    db: Session = Depends(get_db),
):
    """Create a task and optionally initial steps (all pending)."""
    steps_to_create = (
        body.initial_steps if body and body.initial_steps
        else DEFAULT_INITIAL_STEPS
    )
    task = Task(status="pending")
    db.add(task)
    db.flush()
    if body and body.name and body.name.strip():
        task.name = body.name.strip()[:255]
    else:
        task.name = f"未命名任务-{task.id}-{datetime.now().strftime('%m%d%H%M')}"
    for step_key in steps_to_create:
        step = TaskStep(task_id=task.id, step_key=step_key, status="pending")
        db.add(step)
    db.commit()
    db.refresh(task)
    return CreateTaskResponse(id=task.id, name=task.name, status=task.status, created_at=task.created_at)


@router.get("/{task_id}", response_model=TaskDetailResponse)
def get_task(task_id: int, db: Session = Depends(get_db)):
    """Get task by id with steps (ordered by step id)."""
    task = require_task(task_id, db)
    steps = db.query(TaskStep).filter(TaskStep.task_id == task_id).order_by(TaskStep.id).all()
    return TaskDetailResponse(
        id=task.id,
        user_id=task.user_id,
        name=task.name,
        status=task.status,
        created_at=task.created_at,
        updated_at=task.updated_at,
        steps=[TaskStepSchema.model_validate(s) for s in steps],
    )


@router.get("/{task_id}/compare-meta", response_model=CompareMetaResponse)
def get_task_compare_meta(task_id: int, db: Session = Depends(get_db)):
    """Return compare metadata for a task: which items have before/after versions."""
    require_task(task_id, db)
    meta = compute_compare_meta_for_task(task_id, db)
    framework = FrameworkCompareMeta(has_diff=meta["framework_has_diff"])
    chapters_items: list[ChapterCompareMetaItem] = []
    for num in meta["chapter_numbers"]:
        chapters_items.append(
            ChapterCompareMetaItem(number=num, has_diff=True, label=f"第 {num} 章")
        )
    return CompareMetaResponse(
        has_any=meta["has_any"],
        framework=framework,
        chapters=chapters_items,
    )


@router.delete("/{task_id}", status_code=204)
def delete_task(task_id: int, db: Session = Depends(get_db)):
    """Delete a task and its steps. Returns 204 No Content.

    DB records are removed first; uploaded files are cleaned up after a
    successful commit so that a failed commit never leaves orphaned DB rows.
    """
    task = require_task(task_id, db)
    db.query(TaskStep).filter(TaskStep.task_id == task_id).delete()
    db.delete(task)
    db.commit()
    task_dir = config.UPLOAD_DIR / f"task_{task_id}"
    shutil.rmtree(task_dir, ignore_errors=True)
    return None


@router.post("/{task_id}/cancel", status_code=200)
def cancel_task(task_id: int, db: Session = Depends(get_db)):
    """Revoke the currently running Celery step for this task (e.g. one-click cancel)."""
    require_task(task_id, db)
    running = (
        db.query(TaskStep)
        .filter(TaskStep.task_id == task_id, TaskStep.status == "running")
        .first()
    )
    if not running:
        return {"message": "当前无正在执行的步骤", "revoked": False}
    celery_task_id = running.celery_task_id
    revoke_error: str | None = None
    if celery_task_id:
        try:
            celery_app.control.revoke(celery_task_id, terminate=True)
        except Exception as e:
            logger.warning(
                "cancel_task: revoke failed task_id=%s celery_task_id=%s: %s",
                task_id,
                celery_task_id,
                e,
                exc_info=True,
            )
            revoke_error = str(e)[:500]
    running.status = "pending"
    running.celery_task_id = None
    db.commit()
    out: dict = {"message": "已取消当前步骤", "revoked": True, "step_key": running.step_key}
    if revoke_error:
        out["revoke_warning"] = "已向 Broker 发送撤销，但 revoke 调用报错，请查看服务端日志"
    return out


@router.get("", response_model=list[TaskSummary])
def list_tasks(db: Session = Depends(get_db)):
    """List tasks (newest first).

    Uses 3 queries total regardless of task count (avoids 1+2N pattern):
      1. SELECT all tasks
      2. SELECT framework steps WHERE task_id IN (...)
      3. SELECT chapters steps WHERE task_id IN (...)
    """
    tasks = db.query(Task).order_by(Task.created_at.desc()).all()
    if not tasks:
        return []

    task_ids = [t.id for t in tasks]

    framework_steps = (
        db.query(TaskStep)
        .filter(TaskStep.task_id.in_(task_ids), TaskStep.step_key == "framework")
        .all()
    )
    chapters_steps = (
        db.query(TaskStep)
        .filter(TaskStep.task_id.in_(task_ids), TaskStep.step_key == "chapters")
        .all()
    )

    framework_by_task: dict[int, TaskStep] = {s.task_id: s for s in framework_steps}
    chapters_by_task: dict[int, TaskStep] = {s.task_id: s for s in chapters_steps}

    summaries: list[TaskSummary] = []
    for t in tasks:
        meta = compute_compare_meta_from_steps(
            framework_by_task.get(t.id),
            chapters_by_task.get(t.id),
        )
        compare_summary: TaskCompareSummary | None = None
        if meta["has_any"]:
            compare_summary = TaskCompareSummary(
                has_framework=meta["framework_has_diff"],
                chapter_count=len(meta["chapter_numbers"]),
            )
        summaries.append(
            TaskSummary(
                id=t.id,
                name=t.name,
                status=t.status,
                created_at=t.created_at,
                compare_summary=compare_summary,
            )
        )
    return summaries
